File: querydsl-codegen/src/test/java/com/querydsl/codegen/EntitySerializerTest.java
/*
 * Copyright 2015, The Querydsl Team (http://www.querydsl.com/team)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.querydsl.codegen;

import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.io.StringWriter;
import java.sql.Time;
import java.util.Collections;
import java.util.Date;
import java.util.EnumMap;
import java.util.Map;

import org.junit.Test;

import com.querydsl.codegen.utils.JavaWriter;
import com.querydsl.codegen.utils.model.*;
import com.querydsl.core.annotations.PropertyType;

public class EntitySerializerTest {

    private QueryTypeFactory queryTypeFactory = new QueryTypeFactoryImpl("Q", "", "");

    private final TypeMappings typeMappings = new JavaTypeMappings();

    private final EntitySerializer serializer = new DefaultEntitySerializer(typeMappings, Collections.<String>emptySet());

    private final StringWriter writer = new StringWriter();

    public static class Entity {

    }

    @Test
    public void javadocs_for_innerClass() throws IOException {
        EntityType entityType = new EntityType(new ClassType(Entity.class));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("QEntitySerializerTest_Entity is a Querydsl query type for Entity"));
        CompileUtils.assertCompiles("QEntitySerializerTest_Entity", writer.toString());
    }

    @Test
    public void different_package() throws IOException {
        queryTypeFactory = new QueryTypeFactoryImpl("Q", "", ".gen");

        EntityType entityType = new EntityType(new ClassType(Entity.class));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public class QEntitySerializerTest_Entity " +
                "extends EntityPathBase<EntitySerializerTest.Entity>"));
        CompileUtils.assertCompiles("QEntitySerializerTest_Entity", writer.toString());
    }

    @Test
    public void no_package() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type);
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public class QEntity extends EntityPathBase<Entity> {"));
        CompileUtils.assertCompiles("QEntity", writer.toString());
    }

    @Test
    public void original_category() throws IOException {
        Map<TypeCategory, String> categoryToSuperClass
                = new EnumMap<TypeCategory, String>(TypeCategory.class);
        categoryToSuperClass.put(TypeCategory.COMPARABLE, "ComparablePath<Entity>");
        categoryToSuperClass.put(TypeCategory.ENUM, "EnumPath<Entity>");
        categoryToSuperClass.put(TypeCategory.DATE, "DatePath<Entity>");
        categoryToSuperClass.put(TypeCategory.DATETIME, "DateTimePath<Entity>");
        categoryToSuperClass.put(TypeCategory.TIME, "TimePath<Entity>");
        categoryToSuperClass.put(TypeCategory.NUMERIC, "NumberPath<Entity>");
        categoryToSuperClass.put(TypeCategory.STRING, "StringPath");
        categoryToSuperClass.put(TypeCategory.BOOLEAN, "BooleanPath");

        for (Map.Entry<TypeCategory, String> entry : categoryToSuperClass.entrySet()) {
            SimpleType type = new SimpleType(entry.getKey(), "Entity", "", "Entity",false,false);
            EntityType entityType = new EntityType(type);
            typeMappings.register(entityType, queryTypeFactory.create(entityType));

            serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
            assertTrue(entry.toString(), writer.toString().contains("public class QEntity extends " + entry.getValue() + " {"));
        }

    }

    @Test
    public void correct_superclass() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "java.util.Locale", "java.util", "Locale",false,false);
        EntityType entityType = new EntityType(type);
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public class QLocale extends EntityPathBase<Locale> {"));
        CompileUtils.assertCompiles("QLocale", writer.toString());
    }

    @Test
    public void primitive_array() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type);
        entityType.addProperty(new Property(entityType, "bytes", new ClassType(byte[].class)));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public final SimplePath<byte[]> bytes"));
        CompileUtils.assertCompiles("QEntity", writer.toString());
    }

    @Test
    public void include() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type);
        entityType.addProperty(new Property(entityType, "b", new ClassType(TypeCategory.BOOLEAN, Boolean.class)));
        entityType.addProperty(new Property(entityType, "c", new ClassType(TypeCategory.COMPARABLE, String.class)));
        //entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "d", new ClassType(TypeCategory.DATE, Date.class)));
        entityType.addProperty(new Property(entityType, "e", new ClassType(TypeCategory.ENUM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "dt", new ClassType(TypeCategory.DATETIME, Date.class)));
        entityType.addProperty(new Property(entityType, "i", new ClassType(TypeCategory.NUMERIC, Integer.class)));
        entityType.addProperty(new Property(entityType, "s", new ClassType(TypeCategory.STRING, String.class)));
        entityType.addProperty(new Property(entityType, "t", new ClassType(TypeCategory.TIME, Time.class)));

        EntityType subType = new EntityType(new SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2",false,false));
        subType.include(new Supertype(type, entityType));
        typeMappings.register(subType, queryTypeFactory.create(subType));

        serializer.serialize(subType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        CompileUtils.assertCompiles("QEntity2", writer.toString());
    }

    @Test
    public void properties() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type);
        entityType.addProperty(new Property(entityType, "b", new ClassType(TypeCategory.BOOLEAN, Boolean.class)));
        entityType.addProperty(new Property(entityType, "c", new ClassType(TypeCategory.COMPARABLE, String.class)));
        //entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "d", new ClassType(TypeCategory.DATE, Date.class)));
        entityType.addProperty(new Property(entityType, "e", new ClassType(TypeCategory.ENUM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "dt", new ClassType(TypeCategory.DATETIME, Date.class)));
        entityType.addProperty(new Property(entityType, "i", new ClassType(TypeCategory.NUMERIC, Integer.class)));
        entityType.addProperty(new Property(entityType, "s", new ClassType(TypeCategory.STRING, String.class)));
        entityType.addProperty(new Property(entityType, "t", new ClassType(TypeCategory.TIME, Time.class)));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        CompileUtils.assertCompiles("QEntity", writer.toString());
    }

    @Test
    public void superType() throws IOException {
        EntityType superType = new EntityType(new SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2",false,false));
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type, Collections.singleton(new Supertype(superType, superType)));
        typeMappings.register(superType, queryTypeFactory.create(superType));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public final QEntity2 _super = new QEntity2(this);"));
        //CompileUtils.assertCompiles("QEntity", writer.toString());
    }

    @Test
    public void delegates() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType(type);
        Delegate delegate = new Delegate(type, type, "test", Collections.<Parameter>emptyList(), Types.STRING);
        entityType.addDelegate(delegate);
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("return Entity.test(this);"));
        CompileUtils.assertCompiles("QEntity", writer.toString());
    }

    @Test
    public void defaultGeneratedAnnotation() throws IOException {
        EntityType entityType = new EntityType(new ClassType(Entity.class));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        String generatedSourceCode = writer.toString();
        assertTrue(generatedSourceCode.contains(String.format("import %s;", GeneratedAnnotationResolver.resolveDefault().getName())));
        assertTrue(generatedSourceCode.contains("@Generated(\"com.querydsl.codegen.DefaultEntitySerializer\")\npublic class"));
        CompileUtils.assertCompiles("QEntitySerializerTest_Entity", generatedSourceCode);
    }

    @Test
    public void customGeneratedAnnotation() throws IOException {
        EntityType entityType = new EntityType(new ClassType(Entity.class));
        typeMappings.register(entityType, queryTypeFactory.create(entityType));

        new DefaultEntitySerializer(typeMappings, Collections.<String>emptySet(), GeneratedAnnotationResolver.qeryDSLGenerated()).serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        String generatedSourceCode = writer.toString();
        assertTrue(generatedSourceCode.contains("import " + com.querydsl.core.annotations.Generated.class.getName() + ";"));
        assertTrue(generatedSourceCode.contains("@" + com.querydsl.core.annotations.Generated.class.getSimpleName() + "(\"com.querydsl.codegen.DefaultEntitySerializer\")\npublic class"));
        CompileUtils.assertCompiles("QEntitySerializerTest_Entity", generatedSourceCode);
    }

}

File: querydsl-kotlin-codegen/src/test/kotlin/com/querydsl/kotlin/codegen/EntitySerializerTest.kt
/*
 * Copyright 2021, The Querydsl Team (http://www.querydsl.com/team)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.querydsl.kotlin.codegen

import com.querydsl.codegen.Delegate
import com.querydsl.codegen.EntitySerializer
import com.querydsl.codegen.EntityType
import com.querydsl.codegen.GeneratedAnnotationResolver
import com.querydsl.codegen.Property
import com.querydsl.codegen.QueryTypeFactory
import com.querydsl.codegen.QueryTypeFactoryImpl
import com.querydsl.codegen.SimpleSerializerConfig
import com.querydsl.codegen.Supertype
import com.querydsl.codegen.TypeMappings
import com.querydsl.codegen.utils.JavaWriter
import com.querydsl.codegen.utils.model.ClassType
import com.querydsl.codegen.utils.model.SimpleType
import com.querydsl.codegen.utils.model.TypeCategory
import com.querydsl.codegen.utils.model.Types
import com.querydsl.core.annotations.Generated
import com.querydsl.core.annotations.PropertyType
import com.querydsl.kotlin.codegen.CompileUtils.assertCompiles
import org.junit.Assert
import org.junit.Ignore
import org.junit.Test
import java.io.StringWriter
import java.sql.Time
import java.util.*

class EntitySerializerTest {
    private var queryTypeFactory: QueryTypeFactory = QueryTypeFactoryImpl("Q", "", "")
    private val typeMappings: TypeMappings = KotlinTypeMappings()
    private val serializer: EntitySerializer = KotlinEntitySerializer(typeMappings, emptySet())
    private val writer = StringWriter()

    class Entity

    @Test
    fun javadocs_for_innerClass() {
        val entityType = EntityType(ClassType(Entity::class.java))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("QEntitySerializerTest_Entity is a Querydsl query type for Entity"))
        assertCompiles("QEntitySerializerTest_Entity", writer.toString())
    }

    @Test
    fun different_package() {
        queryTypeFactory = QueryTypeFactoryImpl("Q", "", ".gen")
        val entityType = EntityType(ClassType(Entity::class.java))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("class QEntitySerializerTest_Entity : EntityPathBase<EntitySerializerTest.Entity>"))
        assertCompiles("QEntitySerializerTest_Entity", writer.toString())
    }

    @Test
    fun no_package() {
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type)
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("class QEntity : EntityPathBase<Entity> {"))
        assertCompiles("QEntity", writer.toString())
    }

    @Test
    fun original_category() {
        val categoryToSuperClass: MutableMap<TypeCategory, String> = EnumMap(TypeCategory::class.java)
        categoryToSuperClass[TypeCategory.COMPARABLE] = "ComparablePath<Entity>"
        categoryToSuperClass[TypeCategory.ENUM] = "EnumPath<Entity>"
        categoryToSuperClass[TypeCategory.DATE] = "DatePath<Entity>"
        categoryToSuperClass[TypeCategory.DATETIME] = "DateTimePath<Entity>"
        categoryToSuperClass[TypeCategory.TIME] = "TimePath<Entity>"
        categoryToSuperClass[TypeCategory.NUMERIC] = "NumberPath<Entity>"
        categoryToSuperClass[TypeCategory.STRING] = "StringPath"
        categoryToSuperClass[TypeCategory.BOOLEAN] = "BooleanPath"
        for (entry in categoryToSuperClass.entries) {
            val type = SimpleType(entry.key, "Entity", "", "Entity", false, false)
            val entityType = EntityType(type)
            typeMappings.register(entityType, queryTypeFactory.create(entityType))
            serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
            Assert.assertTrue(entry.toString(), writer.toString().contains("class QEntity : " + entry.value + " {"))
        }
    }

    @Test
    fun correct_superclass() {
        val type = SimpleType(TypeCategory.ENTITY, "java.util.Locale", "java.util", "Locale", false, false)
        val entityType = EntityType(type)
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("class QLocale : EntityPathBase<Locale> {"))
        assertCompiles("QLocale", writer.toString())
    }

    @Test
    fun primitive_array() {
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type)
        entityType.addProperty(Property(entityType, "bytes", ClassType(ByteArray::class.java)))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("val bytes: SimplePath<ByteArray>"))
        assertCompiles("QEntity", writer.toString())
    }

    @Test
    fun include() {
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type)
        entityType.addProperty(Property(entityType, "b", ClassType(TypeCategory.BOOLEAN, Boolean::class.java)))
        entityType.addProperty(Property(entityType, "c", ClassType(TypeCategory.COMPARABLE, String::class.java)))
        //entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(Property(entityType, "d", ClassType(TypeCategory.DATE, Date::class.java)))
        entityType.addProperty(Property(entityType, "e", ClassType(TypeCategory.ENUM, PropertyType::class.java)))
        entityType.addProperty(Property(entityType, "dt", ClassType(TypeCategory.DATETIME, Date::class.java)))
        entityType.addProperty(Property(entityType, "i", ClassType(TypeCategory.NUMERIC, Int::class.java)))
        entityType.addProperty(Property(entityType, "s", ClassType(TypeCategory.STRING, String::class.java)))
        entityType.addProperty(Property(entityType, "t", ClassType(TypeCategory.TIME, Time::class.java)))
        val subType = EntityType(SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2", false, false))
        subType.include(Supertype(type, entityType))
        typeMappings.register(subType, queryTypeFactory.create(subType))
        serializer.serialize(subType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        assertCompiles("QEntity2", writer.toString())
    }

    @Test
    fun properties() {
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type)
        entityType.addProperty(Property(entityType, "b", ClassType(TypeCategory.BOOLEAN, Boolean::class.java)))
        entityType.addProperty(Property(entityType, "c", ClassType(TypeCategory.COMPARABLE, String::class.java)))
        //entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(Property(entityType, "d", ClassType(TypeCategory.DATE, Date::class.java)))
        entityType.addProperty(Property(entityType, "e", ClassType(TypeCategory.ENUM, PropertyType::class.java)))
        entityType.addProperty(Property(entityType, "dt", ClassType(TypeCategory.DATETIME, Date::class.java)))
        entityType.addProperty(Property(entityType, "i", ClassType(TypeCategory.NUMERIC, Int::class.java)))
        entityType.addProperty(Property(entityType, "s", ClassType(TypeCategory.STRING, String::class.java)))
        entityType.addProperty(Property(entityType, "t", ClassType(TypeCategory.TIME, Time::class.java)))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        assertCompiles("QEntity", writer.toString())
    }

    @Test
    fun superType() {
        val superType = EntityType(SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2", false, false))
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type, setOf(Supertype(superType, superType)))
        typeMappings.register(superType, queryTypeFactory.create(superType))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("val _super: QEntity2 by lazy {\n    QEntity2(this)\n  }"))
        //CompileUtils.assertCompiles("QEntity", writer.toString());
    }


    @Test
    @Ignore //TODO: Implement delegates. Or document that extensions need to be used instead?
    fun delegates() {
        val type = SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity", false, false)
        val entityType = EntityType(type)
        val delegate = Delegate(type, type, "test", emptyList(), Types.STRING)
        entityType.addDelegate(delegate)
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        Assert.assertTrue(writer.toString().contains("return Entity.test(this)"))
        assertCompiles("QEntity", writer.toString())
    }

    @Test
    fun defaultGeneratedAnnotation() {
        val entityType = EntityType(ClassType(Entity::class.java))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        val generatedSourceCode = writer.toString()
        Assert.assertTrue(generatedSourceCode.contains("import $generatedAnnotationImport"))
        Assert.assertTrue(generatedSourceCode.contains("@Generated(\"com.querydsl.kotlin.codegen.KotlinEntitySerializer\")\npublic class"))
        assertCompiles("QEntitySerializerTest_Entity", generatedSourceCode)
    }

    @Test
    fun customGeneratedAnnotation() {
        val entityType = EntityType(ClassType(Entity::class.java))
        typeMappings.register(entityType, queryTypeFactory.create(entityType))
        KotlinEntitySerializer(typeMappings, emptySet(), Generated::class.java).serialize(entityType, SimpleSerializerConfig.DEFAULT, JavaWriter(writer))
        val generatedSourceCode = writer.toString()
        Assert.assertTrue(generatedSourceCode.contains("import " + Generated::class.java.name))
        Assert.assertTrue(generatedSourceCode.contains("@${Generated::class.java.simpleName}(\"com.querydsl.kotlin.codegen.KotlinEntitySerializer\")\npublic class"))
        assertCompiles("QEntitySerializerTest_Entity", generatedSourceCode)
    }
}